### ISSUE TITLE
Review timeout thresholds

### ISSUE THREAD

**Issue by christian-kirschnick** [corona-warn-app/cwa-server#423]
We need to review the timeout thresholds for all affected components:

- incoming connections (e.g. submission endpoint)
- DeferredResult in submission endpoint
- FeignClient timeouts when calling verification server (+ retry)

**Comment by stevesap** [comment]
**Submission endpoint**

Not sure what default is currently but it appears to be higher than that of `DeferredResult` as `DeferredResult` times out first (on my dev env) and that in turn causes endpoint to immediately return `HTTP 500 Internal Server Error`
 
Can be set at method level using `@Transactional(timeout = X)` (https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/annotation/Transactional.html)

Or can be set via configuration `spring.mvc.async.request-timeout` (https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-application-properties.html#web-properties)

**DeferredResult** 

From Spring Doc: By default not set in which case the default configured in the MVC Java Config or the MVC namespace is used, or if that's not set, then the timeout depends on the default of the underlying server. (https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/context/request/async/DeferredResult.html)

Appears to be 30 seconds currently (on my dev env)

Can be set on object construction as well

**FeignClient**

Can be configured in `application.yml` (https://cloud.spring.io/spring-cloud-static/spring-cloud-openfeign/2.2.3.RELEASE/reference/html/#spring-cloud-feign-overriding-defaults)

Not sure what default is currently but on my dev env it is higher than the `DeferredResult` timeout.

Conclusion (on dev env):
- `FeignClient` timeout > `DeferredResult` timeout AND Submission Endpoint timeout > `DeferredResult` timeout

Recommendation:
- Ensure Submission Endpoint timeout > `DeferredResult` timeout > `FeignClient` timeout

**Comment by RomanNess** [comment]
@jgs88 and me did some investigations on the retry topic today.

We added a configurable delay to a fork of `atakanyenel/cwa-verification-fake` for easier manual testing of timeouts (see [cosee/cwa-verification-fake](https://github.com/cosee/cwa-verification-fake)).

## App timeouts
We did also investigate the timeout values in the mobile apps because they will probably be the base for any timeouts set on the server.

It seems like the Android app has a [read timeout of 10s](https://github.com/corona-warn-app/cwa-app-android/blob/master/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/http/config/HTTPVariables.kt#L22) and a [transaction timeout of 60s](https://github.com/corona-warn-app/cwa-app-android/blob/c3416440e644a918f4001489de85f203c7681099/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/TimeVariables.kt#L37).

In our understanding, the iOS app uses the [default timeout of 60s](https://developer.apple.com/documentation/foundation/urlrequest/2011457-init) which [is no overwritten on call](https://github.com/corona-warn-app/cwa-app-ios/blob/a6b885f42f9444d890210a4f59d62041ca0c834d/src/xcode/ENA/ENA/Source/Client/HTTP%20Client/HTTPClient.swift#L416).

In our opinion, a read timeout of 10s with retries and exponential backoff could be a valid solution for both mobile apps.

## Submission endpoint
In the current implementation the submission endpoint still returns a `DeferredResult` but it is implemented [synchronous](https://github.com/corona-warn-app/cwa-server/blob/97331e3a4886cf9737c03d0268216274b9ac9b14/services/submission/src/main/java/app/coronawarn/server/services/submission/controller/SubmissionController.java#L91-L111). Since it is not running asynchronously, neither the property `spring.mvc.async.request-timeout` nor the `@Transactional(timeout = X)` annotation will lead to the expected behavior.

`spring.mvc.async.request-timeout` should have no effect at all on synchronous endpoints.
`@Transactional(timeout = X)` will rollback the transaction when the annotated method takes longer than the specified timeout in seconds, but it will not interrupt synchronous code (e.g. a call to another service).

A possible solution is to run the method that handles real submission requests in a suitable async executor again. Then both the `spring.mvc.async.request-timeout` property and a timeout on the `DeferredResult` seems to lead to the same timeout behavior. On timeout the asynchronous execution is interrupted and the method throws an `AsyncRequestTimeoutException`.

## Database Transactions
`@Transactional(timeout = X)` can be used on DAO Layer where the code uses practically all of it's time in the interaction with the DB. `propagation=REQUIRES_NEW` may be necessary if an outer method is already transactional. It will set the session timeout in the `JpaTransactionManager` and should cancel a DB query when a timeout is triggered. We have not used this feature in production yet, but we could see it working to recover an overburdened DB if the client has implemented exponential backoff.

**Comment by stevesap** [comment]
Thanks very much for your detailed analysis. 

> A possible solution is to run the method that handles real submission requests in a suitable async executor again

I agree this could be a solution. However, as described in #518, we decided that it only makes sense for us to use an executor for fake request handling and as of now, this approach is working for us.

So did you by any chance also experiment with setting timeout on the submission endpoint using the current implementation? Perhaps using `server.tomcat.connection-timeout`?

**Comment by RomanNess** [comment]
> So did you by any chance also experiment with setting timeout on the submission endpoint using the current implementation? Perhaps using server.tomcat.connection-timeout?

I did just experiment with `server.tomcat.connection-timeout` and the deprecated `server.connection-timeout`. None of them seems to have an effect on long running requests.

In my opinion this is not surprising since the [Tomcat documentation](https://tomcat.apache.org/tomcat-8.5-doc/config/http.html) states:
> The number of milliseconds this Connector will wait, after accepting a connection, for the request URI line to be presented.

I was only able to trigger the timeout with a `nc` call that opens a connection on the port but does not post a request.
```
$ time nc localhost 8080 -vv 
Connection to localhost port 8080 [tcp/http-alt] succeeded!
nc localhost 8080 -vv  0.01s user 0.01s system 0% cpu 2.018 total
```
--- 
> I agree this could be a solution. However, as described in #518, we decided that it only makes sense for us to use an executor for fake request handling and as of now, this approach is working for us.

I can understand that you hesitate to introduce asynchronous handling in the submission endpoint again after you just got rid of it. However, in my understanding the main issue in #518 was that it used the ForkJoinPool instead of a more suitable executor.

A downside of the synchronous request handler is that I don't know of a built-in configuration property in Spring Boot to cancel long running requests. This can lead to issues if requests timeout on the mobile client but are still active on the server. Though, every time I have seen this happening in real-life, it was caused by an overloaded DB and fake requests do not reach the DB.

We should also keep in mind that non-fake submission requests are very rare compared to fake requests. The only case where a timeout might be useful are probably submission requests with an invalid Tan sent by attackers or bored users. Calls to the VerificationServer (that are the second possible source of long running requests) can be cancelled by a timeout on the FeignClient as proposed in #581.

Considering all of this my suggestion is to either 1) keep the submission endpoint synchronous and to go on without timeouts or 2) make the endpoint asynchronous again with a suitable executor and set a timeout on the `DeferredResult`. 
Basically you trade simple synchronous code that cannot be interrupted in the worst case against more complex asynchronous code that might cause issues itself in another worst case.
